csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedExpr.qll

```diff
@@ -1918,3 +1963,91 @@ class TranslatedDelegateCall extends TranslatedNonConstantExpr {
     result = DelegateElements::getInvoke(expr)
   }
 }
+
+/**
+ * Represents the IR translation of creation expression. Can be the translation of an
+ * `ObjectCreation` or a `DelegateCreation`.
+ * The `NewObj` instruction denotes the fact that during initialization a new
+ * object is allocated, which is then initialized by the constructor.
+ */
+abstract class TranslatedCreation extends TranslatedCoreExpr, TTranslatedCreationExpr,
+  ConstructorCallContext {
+  TranslatedCreation() { this = TTranslatedCreationExpr(expr) }
+
+  override TranslatedElement getChild(int id) {
+    id = 0 and result = this.getConstructorCall()
+    or
+    id = 1 and result = this.getInitializerExpr()
+  }
+
+  override predicate hasInstruction(
+    Opcode opcode, InstructionTag tag, Type resultType, boolean isLValue
+  ) {
+    // Instruction that allocated space for a new object,
+    // and returns its address
+    tag = NewObjTag() and
+    opcode instanceof Opcode::NewObj and
+    resultType = expr.getType() and
+    isLValue = false
+  }
+
+  final override Instruction getFirstInstruction() { result = this.getInstruction(NewObjTag()) }
+
+  override Instruction getResult() { result = getInstruction(NewObjTag()) }
+
+  override Instruction getReceiver() { result = getInstruction(NewObjTag()) }
+
+  override Instruction getInstructionSuccessor(InstructionTag tag, EdgeKind kind) {
+    kind instanceof GotoEdge and
```

Comment:
Why a `GOTO`?

Reply:
Because the constructor call happens after the new object is allocated and we have a direct edge. `Goto` edges here mean successor edges, maybe that's what's causing the confusion?

Reply:
Ah, I see. I thought it was related to `goto` statements. Would `SuccessorEdge` or `NormalEdge` perhaps be a better name, to avoid confusion?

Reply:
That would probably make it clearer, but the file that contains the edge kinds, `EdgeKind.qll` is a shared file, so I would have to talk to the cpp team. Should I do this for this PR?

Reply:
No, fine to do in a follow-up PR.